NullPointerException When Calling Graph API Concurrently

We are using microsoft-graph version 1.0.0 jar for making calls to the microsoft graph api. While the api calls work fine when executing in a single thread, very frequently we do get the below exception which breaks the http connection with microsoft graph servers and any further attempt to make api calls fails.

We have a pool of 5 threads that call the microsoft graph api on the "teams" resource to get channel conversations and their replies. The later is a batch request. The threads do handle "429 - Too many requests" error by waiting for "Reply-After" interval before making any further calls. During the code execution we could see that the below exception comes after we receive few such api rate limiting errors. 

Any points for further troubleshooting would be helpful. Thanks.

The exception is 

```
0420 12:52:57.817+0530 {/} ERROR [pool-3-thread-3][c.e.o.t.s.MicrosoftTeamsService-341 ] calling microsoft graph: 19:b08b3951a7654111add8d6b1c4455663@thread.skype error: {} com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:356)
	at com.microsoft.graph.http.DefaultHttpProvider.send(DefaultHttpProvider.java:197)
	at com.microsoft.graph.http.DefaultHttpProvider.send(DefaultHttpProvider.java:177)
	at com.microsoft.graph.http.BaseRequest.send(BaseRequest.java:277)
	at com.microsoft.graph.http.CustomRequest.post(CustomRequest.java:106)
	at com.example.officeintegration.teams.service.MicrosoftTeamsService.executeBatchHttpPostRequest(MicrosoftTeamsService.java:612)
Caused by: java.lang.NullPointerException: source
	at java.util.Objects.requireNonNull(Objects.java:228)
	at java.util.Scanner.<init>(Scanner.java:578)
	at com.microsoft.graph.http.DefaultHttpProvider.streamToString(DefaultHttpProvider.java:440)
	at com.microsoft.graph.http.GraphServiceException.createFromConnection(GraphServiceException.java:305)
	at com.microsoft.graph.http.DefaultHttpProvider.handleErrorResponse(DefaultHttpProvider.java:376)
	at com.microsoft.graph.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:309)
	... 208 common frames omitted
```

--- a/src/main/java/com/microsoft/graph/http/GraphServiceException.java
+++ b/src/main/java/com/microsoft/graph/http/GraphServiceException.java
@@ -311,7 +311,10 @@ public class GraphServiceException extends ClientException {
         }
 
         final String responseMessage = connection.getResponseMessage();
-        final String rawOutput = DefaultHttpProvider.streamToString(connection.getInputStream());
+        String rawOutput = "{}";
+        if(connection.getInputStream() != null) {
+        	rawOutput = DefaultHttpProvider.streamToString(connection.getInputStream());
+        }
         GraphErrorResponse error;
         try {
             error = serializer.deserializeObject(rawOutput, GraphErrorResponse.class, connection.getResponseHeaders());

--- a/src/test/java/com/microsoft/graph/http/GraphServiceExceptionTests.java
+++ b/src/test/java/com/microsoft/graph/http/GraphServiceExceptionTests.java
@@ -4,7 +4,9 @@ import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
+import java.io.InputStream;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Map;
@@ -68,6 +70,48 @@ public class GraphServiceExceptionTests {
         };
         try{
             exception = GraphServiceException.createFromConnection(new MockHttpRequest(),null,null,new MockConnection(data){},logger);
+            success = true;
+        }catch (IOException ex){
+            failure = true;
+        }
+
+        assertTrue(success);
+        assertFalse(failure);
+
+        String message = exception.getMessage();
+        assertTrue(message.indexOf("Error code: Unable to parse error response message") == 0);
+        assertTrue(message.indexOf("http://localhost") > 0);
+    }
+	
+	@Test
+	public void testNullConnection() {
+		DefaultLogger logger = new DefaultLogger();
+        GraphServiceException exception = null;
+        Boolean success = false;
+        Boolean failure = false;
+        final ITestConnectionData data = new ITestConnectionData() {
+            @Override
+            public int getRequestCode() {
+                return 401;
+            }
+
+            @Override
+            public String getJsonResponse() {
+                return "{}";
+            }
+
+            @Override
+            public Map<String, String> getHeaders() {
+                return new HashMap<>();
+            }
+        };
+        try{
+            exception = GraphServiceException.createFromConnection(new MockHttpRequest(),null,null,new MockConnection(data) {
+            	@Override
+            	public InputStream getInputStream() throws IOException {
+                    return null;
+                }
+            },logger);
             success = true;
         }catch (IOException ex){
             failure = true;

